executor/src/witgen/vm_processor.rs

```diff
@@ -383,18 +450,117 @@ impl<'a, T: FieldElement> VmProcessor<'a, T> {
+
+        let row_pair = RowPair::new(
+            &self.data[row_index as usize],
+            &self.data[row_index as usize + 1],
+            (row_index + self.row_offset) % self.fixed_data.degree,
+            self.fixed_data,
+            UnknownStrategy::Unknown,
+        );
+
+        let mut identity_processor = IdentityProcessor::new(self.fixed_data, mutable_state);
+        let updates = identity_processor
+            .process_link(left, right, &row_pair)
+            .map_err(|e| {
+                log::warn!("Error in outer query: {e}");
+                log::warn!("Some of the following entries could not be matched:");
+                for (l, r) in left.iter().zip(right.expressions.iter()) {
+                    if let Ok(r) = row_pair.evaluate(r) {
+                        log::warn!("  => {} = {}", l, r);
+                    }
+                }
+                e
+            })?;
+
+        let progress = self.apply_updates(row_index, &updates, || "outer query".to_string());
+
+        let outer_assignments = updates
+            .constraints
+            .into_iter()
+            .filter(|(poly, _)| !self.witnesses.contains(&poly.poly_id()))
+            .collect::<Vec<_>>();
+
+        Ok((progress, outer_assignments))
+    }
+
+    /// Sets the inputs to the values given in [VmProcessor::inputs] if they are not already set.
+    /// Typically, inputs will have a constraint of the form: `((1 - instr__reset) * (_input' - _input)) = 0;`
+    /// So, once the value of `_input` is set, this function will do nothing until the next reset instruction.
+    /// However, if `_input` does become unconstrained, we need to undo all changes we've done so far.
+    /// For this reason, we keep track of all changes we've done to inputs in [VmProcessor::previously_set_inputs].
+    fn set_inputs_if_unset(&mut self, row_index: DegreeType) -> bool {
+        let mut input_updates = EvalValue::complete(vec![]);
+        for (poly_id, value) in self.inputs.iter() {
+            match &self.data[row_index as usize][poly_id].value {
+                CellValue::Known(_) => {}
+                CellValue::RangeConstraint(_) | CellValue::Unknown => {
+                    input_updates.combine(EvalValue::complete([(
+                        &self.fixed_data.witness_cols[poly_id].poly,
+                        Constraint::Assignment(*value),
+                    )]));
+                }
+            };
+        }
+
+        for (poly, _) in &input_updates.constraints {
+            let poly_id = poly.poly_id();
+            if let Some(start_row) = self.previously_set_inputs.remove(&poly_id) {
+                log::trace!(
+                    "    Resetting previously set inputs for column: {}",
+                    self.fixed_data.column_name(&poly_id)
+                );
+                for row_index in start_row..row_index {
+                    self.data[row_index as usize][&poly_id].value = CellValue::Unknown;
```

Comment:
What happens to those unknown values later on? We don't reset the row number, do we?

Reply:
Yeah, it's a bit weird, because in practice, these values are going to stay unknown and are eventually set to 0.

The point is that they should be equal to the inputs of the *previous* block. Either we know those already (because `VmProcessor` gets initialized with a first row that has these values, but then we never reset), or we are in the first block, in which case the previous block is the last block, which is the infinite loop. But the infinite look does not have inputs, so it's fine if the values are unknown.

If the last block *did* have inputs, we'd have to process the first block again, to see how far we have to propagate these inputs. This reset strategy could stay the same though.